text/0521-find-by-identifier.md

```diff
@@ -0,0 +1,167 @@
+would error.
+
+This is because there is no meaningful way for the store to create a
+network request for a record that does not exist yet (a
+`NewResourceIdentifierObject`), or for one for which the only thing
+we know is a locally assigned id (`Identifier`).
+
+Finally, we also deprecate `hasRecordForId` in favor of `peekRecord`,
+the private `setRecordId` (usage of this feature should be done through
+the store-wrapper instead), and the private `recordForId` for which the replacement is based on use-case.
+
+```ts
+interface Identifier {
+  lid: string;
+}
+
+export interface ExistingResourceIdentifierObject {
+  id: string;
+  type: string;
+  lid?: string;
+}
+
+export interface NewResourceIdentifierObject {
+  id: string | null;
+  type: string;
+  lid: string;
+}
+
+export type Peekable =
+  | Identifier
+  | ExistingResourceIdentifierObject
+  | NewResourceIdentifierObject;
+
+class Store {
+  /*
+  @deprecated
+  */
+  peekRecord(type: string, id: string | number): Record | null {}
+
+  // replacement
+  peekRecord(identifier: Peekable): Record | null {}
+
+  /*
+  @deprecated
+  */
+  findRecord(type: string, id: string | number): PromiseRecord {}
+
+  // replacement
+  findRecord(identifier: Peekable): Promise<Record> {}
+
+  /*
+  @deprecated
+  */
+  getReference(modelName: string, id: string | number): RecordReference {}
+
+  // replacement
+  getReference(identifier: Peekable): RecordReference {}
+
+  /*
+  @deprecated
+
+  replacement is `peekRecord() !== null`
+  */
+  hasRecordForId(modelName: string, id: string | number): boolean {}
+
+  /*
+    @private
+    @deprecated
+
+    replacement is peekRecord, createRecord, or push based on use-case
+  */
+  recordForId(modelName: string, id: string | number): Record {}
+
+ /*
+    @private
+    @deprecated
+
+    replacement is to use the appropriate StoreWrapper method
+  */
+  setRecordId(modelName: string, newId: string, clientId: string) {
+}
+```
+
+## How we teach this
+
+The general migration is fairly simple. For instance, imagine a Route that both peeks a record and makes a findRecord request.
+
+**before**
+
+```ts
+{
+  @service('store') store;
+
+  model({ user_id, post_id }) {
+    let user = this.store.peekRecord('user', user_id);
+
+    return hash({
+        user,
+        post: this.store.findRecord('post', post_id)
+    });
+  }
+}
+```
+
+**after**
+
+```ts
+{
+  @service('store') store;
+
+  model({ user_id, post_id }) {
+    let user = this.store.peekRecord({ type: 'user', id: user_id });
+
+    return hash({
+        user,
+        post: this.store.findRecord({ type: 'post', id: post_id })
+    });
+  }
+}
+```
+
+API Documentation and Guides would need updated to use these new method
+signatures and to discuss how these APIs fit with Identifiers.
+
+## Drawbacks
+
+- It changes the function signatures of two of the most commonly used
+  EmberData APIs. However it also aligns those APIs better with more
+  recently designed APIs.
```

Comment:
Given that we're considering updating two quite common EmberData apis, do we envision releasing a code mod to help make this change mechanical?

Reply:
we could at such time as we deprecate the existing signature, but since we have decided not to this isn't something we need now